music_assistant/server/providers/tidal/__init__.py

```diff
@@ -0,0 +1,394 @@
+    """
+    Return Config entries to setup this provider.
+
+    instance_id: id of an existing provider instance (None if new instance setup).
+    action: [optional] action key called from config entries UI.
+    values: the (intermediate) raw values for config entries sent with the action.
+    """
+
+    # config flow auth action/step (authenticate button clicked)
+    if action == CONF_ACTION_AUTH:
+        async with AuthenticationHelper(mass, values["session_id"]) as auth_helper:
+            TIDAL_SESSION = await tidal_code_login(auth_helper)
+            if not TIDAL_SESSION.check_login():
+                raise LoginFailed("Authentication to Tidal failed")
+            # set the retrieved token on the values object to pass along
+            values[CONF_AUTH_TOKEN] = TIDAL_SESSION.access_token
+            values[CONF_REFRESH_TOKEN] = TIDAL_SESSION.refresh_token
+            values[CONF_EXPIRY_TIME] = TIDAL_SESSION.expiry_time.isoformat()
+            values[CONF_USER_ID] = str(TIDAL_SESSION.user.id)
+
+    # return the collected config entries
+    return (
+        ConfigEntry(
+            key=CONF_AUTH_TOKEN,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Authentication token for Tidal",
+            description="You need to link Music Assistant to your Tidal account.",
+            action=CONF_ACTION_AUTH,
+            action_label="Authenticate on Tidal.com",
+            value=values.get(CONF_AUTH_TOKEN) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_REFRESH_TOKEN,
+            type=ConfigEntryType.SECURE_STRING,
+            label="Refresh token for Tidal",
+            description="You need to link Music Assistant to your Tidal account.",
+            action_label="Authenticate on Tidal.com",
+            value=values.get(CONF_REFRESH_TOKEN) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_EXPIRY_TIME,
+            type=ConfigEntryType.STRING,
+            label="Expiry time of auth token for Tidal",
+            action_label="Authenticate on Tidal.com",
+            value=values.get(CONF_EXPIRY_TIME) if values else None,
+        ),
+        ConfigEntry(
+            key=CONF_USER_ID,
+            type=ConfigEntryType.STRING,
+            label="Your Tidal User ID",
+            description="This is your unique Tidal user ID.",
+            action_label="Authenticate on Tidal.com",
+            value=values.get(CONF_USER_ID) if values else None,
+        ),
+    )
+
+
+class TidalProvider(MusicProvider):
+    """Implementation of a Tidal MusicProvider."""
+
+    async def handle_setup(self) -> None:
+        """Handle async initialization of the provider."""
+        self._cache_dir = CACHE_DIR
+        self._tidal_session = TIDAL_SESSION
+        self._tidal_user_id = self.config.get_value(CONF_USER_ID)
+        # try to get a token, raise if that fails
+        self._cache_dir = os.path.join(CACHE_DIR, self.instance_id)
```

Comment:
is the CACHE_DIR some leftover old code ?

Reply:
Yes it is indeed